README.md
[![Build Status](https://travis-ci.org/Boereck/fn_block.svg?branch=master)](https://travis-ci.org/Boereck/fn_block) 
[![AppVeyor Status](https://ci.appveyor.com/api/projects/status/apctgp7w8qcwttag?svg=true)](https://ci.appveyor.com/project/Boereck/fn-block) 
[![Crates.io Version](https://img.shields.io/crates/v/fn_block.svg)](https://crates.io/crates/fn_block)

# `fn_block` Crate

Library defining macros for calling blocks or expressions in a closure.

## Quick Introduction

This library was mostly written to allow "safe navigation" with the `?.` operator combination
(seemingly) without jumping out of the current function. This allows a similar use of the operator
as in other languages (such as Swift, C# or Kotlin).

To use this library, you have to add it to the dependencies of your `Cargo.toml` file

```toml
[dependencies]
fn_block = "0.2.1"
```

Then add the following lines to your module:

```rust
#[macro_use]
extern crate fn_block;
use fn_block::*;
```
Instead of the wildcard, you can also just import the symbols you need. 

Here is an example on how to use the crate:

```rust
let o = Some("Foobar");
let s = fn_expr!{ o?.get(0..3)?.to_lowercase().into_some() };
assert_eq!("foo", s.unwrap());
```

Please visit the [API Documentation](https://docs.rs/fn_block/latest/) for more details.

## Functionality Overview

In short, this crate provides the following APIs:

* The [`fn_expr`] macro allows wrapping an expression into a lambda that is directly called.
* The [`IntoSome`] trait, which is implemented for all `Sized` types, allows to call [`into_some`] 
  on a value to move it into an `Option::Some`.
* The [`IntoOk`] trait, which is implemented for all `Sized` types, allows to call [`into_ok`] 
  on a value to move it into an `Result::Ok`.

For more examples, please have a look at the test module.

## Unstable Features

To use unstable features, the dependency declaration in your `Cargo.toml` has to be updated:

```toml
[dependencies]
fn_block = { version = "0.2.1", features = ["unproven"] }
```
Note that this crate's unstable features *do* work on stable Rust.

The following unstable APIs are available:

* The [`fn_try`] macro allows wrapping an expression into a lambda, being called directly and recover from errors directly afterwards.

## License

The fn_block crate is licensed under the following licenses:

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) / http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) / http://opensource.org/licenses/MIT)

Choose under which you want to use the library.

[`fn_expr`]: https://docs.rs/fn_block/latest/fn_block/macro.fn_expr.html
[`fn_block`]: https://docs.rs/fn_block/latest/fn_block/macro.fn_block.html
[`fn_try`]: https://docs.rs/fn_block/latest/fn_block/macro.fn_try.html
[`IntoSome`]: https://docs.rs/fn_block/latest/fn_block/trait.IntoSome.html
[`into_some`]: https://docs.rs/fn_block/latest/fn_block/trait.IntoSome.html#tymethod.into_some
[`IntoOk`]: https://docs.rs/fn_block/latest/fn_block/trait.IntoOk.html
[`into_ok`]: https://docs.rs/fn_block/latest/fn_block/trait.IntoOk.html#tymethod.into_ok

CHANGES.md
# `fn_block` Version History

## 0.1.0 (2018-05-25)

* Added `fn_expr` macro
* Added `fn_block` macro
* Added `IntoOk` trait and implementation for every type
* Added `IntoSome` trait and implementation for every type

## 0.2.0 (2018-07-04)

* Added unstable/unproven marco `fn_try` (crate feature `unproven` needs to be enabled)
* Deprecated `fn_block` macro, since `fn_expr` can also wrap blocks. This is a bit ironic, since the crate name is `fn_block`

## 0.2.1 (2018-07-04)

* Only crate `Cargo.toml` metadata fixes

src/tests.rs
use super::*;

/// Trait implemented for all sized types, providing a `when` function,
/// wrapping the element it is called on into an `Optional::Some` if the
/// given predicate holds true for the value and `Optional::None` otherwise.
trait Optionalize: Sized {
    fn when<F>(self, predicate: F) -> Option<Self>
    where
        F: FnOnce(&Self) -> bool,
    {
        if predicate(&self) {
            Some(self)
        } else {
            None
        }
    }
}

/// Implementation of Optionalize for all values
impl<T> Optionalize for T {}

#[test]
fn fn_block_some() {
    let o: Option<i32> = Some(42);
    let foo = fn_block!{{
        let gt_zero = o?.when(|&i| i > 0);
        gt_zero?.when(|&i| i%2 == 0)
    }};
    assert_eq!(42, foo.expect("result"));
}

#[test]
fn fn_block_none() {
    let o: Option<i32> = Some(42);
    let foo = fn_block!{{
        let big_num = o?.when(|&i| i > 1000);
        big_num?.when(|&i| i < 2000)
    }};
    assert!(foo.is_none());
}

#[test]
fn fn_expr_block_some() {
    let o: Option<i32> = Some(42);
    let foo = fn_expr!{{
        let gt_zero = o?.when(|&i| i > 0);
        gt_zero?.when(|&i| i%2 == 0)
    }};
    assert_eq!(42, foo.expect("result"));
}

#[test]
fn fn_expr_block_none() {
    let o: Option<i32> = Some(42);
    let foo = fn_expr!{{
        let big_num = o?.when(|&i| i > 1000);
        big_num?.when(|&i| i < 2000)
    }};
    assert!(foo.is_none());
}

#[test]
fn fn_expr_some() {
    let o: Option<i32> = Some(42);
    let foo = fn_expr!{ o?.when(|&i| i > 0)?.when(|&i| i%2 == 0) };
    assert_eq!(42, foo.expect("result"));
}

#[test]
fn fn_expr_none() {
    let o: Option<i32> = Some(42);
    let foo = fn_expr!{ o?.when(|&i| i > 1000)?.when(|&i| i < 2000) };
    assert!(foo.is_none());
}

#[test]
fn fn_block_resulttype() {
    use std::num::ParseIntError;
    let res = fn_block!{ Result<u32,ParseIntError>: {
        let s = "4711";
        s.parse()
    }};
    assert_eq!(4711, res.unwrap());
}

#[test]
fn fn_expr_resulttype() {
    use std::num::ParseIntError;
    let res = fn_expr!{ Result<u32,ParseIntError>: "4711".parse() };
    assert_eq!(4711, res.unwrap());
}

#[test]
fn into_ok() {
    let r: Result<&str, u16> = "foo".into_ok();
    assert_eq!("foo", r.expect("result is Ok"));
}

#[test]
fn into_some() {
    let r: Option<u32> = 42.into_some();
    assert_eq!(42, r.expect("result is Some"));
}

#[test]
fn showcase() {
    // Not an actual test, but a showcase for several alternative ways to express the same
    let o = Some("Foobar");
    let s = o.and_then(|st| st.get(0..3)).map(|st| st.to_lowercase());
    assert_eq!("foo", s.expect("result is Some"));

    let s = (|| Some(o?.get(0..3)?.to_lowercase()))();
    assert_eq!("foo", s.expect("result is Some"));

    use super::IntoSome;
    let s = fn_expr!{ o?.get(0..3)?.to_lowercase().into_some() };
    assert_eq!("foo", s.expect("result is Some"));
}

use std::num::ParseIntError;
use std::str::Utf8Error;

enum ConvertErr {
    StrParseErr,
    IntParseErr,
}

impl From<Utf8Error> for ConvertErr {
    fn from(_: Utf8Error) -> ConvertErr {
        ConvertErr::StrParseErr
    }
}
impl From<ParseIntError> for ConvertErr {
    fn from(_: ParseIntError) -> ConvertErr {
        ConvertErr::IntParseErr
    }
}

#[test]
#[cfg(feature = "unproven")]
fn fn_catch_result() {
    use std::str::from_utf8;

    let s: &[u8] = &[0x0020, 0x0034, 0x0032];
    let i = fn_try! {
        from_utf8(s)?.trim().parse::<u32>()?
        => catch {
            ConvertErr::StrParseErr => 0u32,
            ConvertErr::IntParseErr => u32::max_value()
        }
    };
    assert_eq!(42, i);
}

#[test]
#[cfg(feature = "unproven")]
fn fn_catch_error() {
    use std::str::from_utf8;

    let s: &[u8] = &[0x0020, 0x005A, 0x0032];
    let i = fn_try! {
        from_utf8(s)?.trim().parse::<u32>()?
        => catch {
            ConvertErr::StrParseErr => 0u32,
            ConvertErr::IntParseErr => u32::max_value()
        }
    };
    assert_eq!(u32::max_value(), i);
}


benches/fn_block_benchmark.rs
#[macro_use]
extern crate criterion;
extern crate fn_block;

#[macro_use]
use fn_block::*;
use criterion::{Criterion, Fun};

fn pure_option(o: Option<&str>) {
    let s = o.and_then(|st| st.get(0..3)).map(|st| st.to_lowercase());
}

fn closure(o: Option<&str>) {
    let s = (|| Some(o?.get(0..3)?.to_lowercase()))();
}

fn fn_expr_usage(o: Option<&str>) {
    let s = fn_expr!{ o?.get(0..3)?.to_lowercase().into_some() };
}

fn bench_option_combinators(c: &mut Criterion) {
    let fn_expr_fn = Fun::new("fn_expr Wrapper", |b, o| b.iter(|| fn_expr_usage(*o)));
    let closure_fn = Fun::new("Closure Wrapper", |b, o| b.iter(|| closure(*o)));
    let pure_opts_fn = Fun::new("Pure Options", |b, o| b.iter(|| pure_option(*o)));

    let functions = vec![pure_opts_fn, closure_fn, fn_expr_fn];
    let opt = Some("Foobar");

    c.bench_functions("Option combinations", functions, opt);
}

criterion_group! {
    name = benches;
    config = Criterion::default().sample_size(500);
    targets = bench_option_combinators
}

criterion_main!(benches);


src/lib.rs
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! # Library for wrapping a block or expression into a closure and call it
//!
//! ## Introduction
//!
//! The functionality of this crate was defined, because every now and then it is unpractical,
//! that the `?` operator in rust has a "unwrap or return" semantic. In many other languges
//! (such as Swift, C#, Kotlin, etc.) the `?.` operator allows "safe navigation",
//! and this would be practical for the `Option` and `Result` type in Rust every now and then.
//! The navigation code could be extracted to another function,
//! but oftentimes this would fragment code that actually belongs together.
//! Another workaround is to wrap a `?.` call chain into a closure and
//! call it directly. For example, consider the following piece of code:
//!
//! ```rust
//! let o = Some("Foobar");
//! let s = o.and_then(|st| st.get(0..3)).map(|st| st.to_lowercase());
//! assert_eq!("foo", s.unwrap());
//! ```
//!
//! The second line would be more readable using the `?.` navigation. But this would lead
//! to a return from the entire function if an empty option is being accessed. So the call
//! chain can be wrapped into a cosure which is then immediately called. The following code
//! shows, how the previous example can be written in this style:
//!
//! ```rust
//! let o = Some("Foobar");
//! let s = (|| Some(o?.get(0..3)?.to_lowercase()) )();
//! assert_eq!("foo", s.unwrap());
//! ```
//!
//! While this code is very terse, it is not necessarily the easiest to read.
//! The `fn_block` crate provides functionality to make the code above
//! a little easier to read:
//!
//! ```rust
//! # #[macro_use]
//! # use fn_block::*;
//! let o = Some("Foobar");
//! let s = fn_expr!{ o?.get(0..3)?.to_lowercase().into_some() };
//! assert_eq!("foo", s.unwrap());
//! ```
//!
//! It was considered to provide a version of the macros that automatically
//! calls `into()` on the result of the expression/block. This would allow an automatic
//! conversion of a value to the actual return type, provided a fitting implementation
//! of the `Into` trait was in scope. But this was considered too implicit. The `?` operator
//! already performs implicit conversion of error types.
//!
//!
//! # Unstable features
//!
//! To enable unstable features, the crate feature `unproven` must be enabled
//! explicitly. Here is an example dependency declaration that can be added to
//! a users `Cargo.toml` file to enable the unstable features:
//!
//! ```toml
//! [dependencies]
//! fn_block = { version = "0.2.0", features = ["unproven"] }
//! ```
//!
//! Note that this crate's unstable features *do* work on stable Rust.
//!
//! The unstable macro [`fn_try!`] does call an expression in a lambda and *does* wrap
//! the sucess value into a `Result::Ok`. It then enforces to recover from the error type
//! in a following `=> catch` block. The reasons behind this descision is documented in
//! the [`fn_try!`] documentation.
//! Overly simple example usage:
//! ```rust
//! # #[macro_use]
//! # use fn_block::*;
//! use std::str::from_utf8;
//! use std::error::Error;
//! struct ConvertErr();
//! impl <T: Error> From<T> for ConvertErr {
//! 	fn from(_: T) -> ConvertErr {ConvertErr()}
//! }
//! let bytes : &[u8] = &[0x0020,0x0034,0x0032];
//! let res_int = fn_try!{
//! 	from_utf8(bytes)?.trim().parse::<u32>()?
//! 	=> catch {
//! 		ConvertErr() => 0u32
//! 	}
//! };
//! assert_eq!(res_int, 42);
//! ```
//! A more verbose and realistic version of the example above is available in
//! the [`fn_try!`] documentation.
//!
//! [`fn_try!`]: macro.fn_try.html

///////////////////////
// Macro definitions //
///////////////////////

/// *NOTE*: This macro has been deprecated! Use `fn_expr` instead
///
/// # Intro
///
/// This macro wraps a given rust code block into a closure and
/// directly calls the closure. Optionally the return type of the
/// closure can be specified first and separeted with a colon from
/// the body block.
///
/// # Example without return type:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// let o = Some("Foobar");
/// let s = fn_expr!{ o?.get(0..3)?.to_lowercase().into_some() };
/// assert_eq!("foo", s.unwrap());
/// ```
///
/// # Example with return type:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// use std::str::from_utf8;
/// use std::error::Error;
/// struct ConvertErr();
/// impl <T: Error> From<T> for ConvertErr {
/// 	fn from(_: T) -> ConvertErr {ConvertErr()}
/// }
/// let bytes : &[u8] = &[0x0020,0x0034,0x0032];
/// let res_int = fn_block!{Result<u32,ConvertErr>: {
/// 	let str = from_utf8(bytes)?.trim();
/// 	str.parse::<u32>()?.into_ok()
/// }}.unwrap_or(0u32);
/// assert_eq!(res_int, 42);
/// ```
///
/// Note that the examples use the traits [`IntoSome`] and [`IntoOk`],
/// Defined in this crate.
///
/// [`IntoSome`]: trait.IntoSome.html
/// [`IntoOk`]: trait.IntoOk.html
#[macro_export]
#[deprecated(
    since = "0.2.0",
    note = "Please use `fn_expr` instead, since it can also be used to wrap blocks. Blocks are expressions as well."
)]
macro_rules! fn_block {
    ($return_type:ty : $body:block) => {
        (|| -> $return_type { $body })()
    };
    ($body:block) => {
        (|| $body)()
    };
}

/// This macro wraps a given rust code block into a closure and
/// directly calls the closure. Optionally the return type of the
/// closure can be specified first and separeted with a colon from
/// the body expression.
///
/// # Example without return type:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// let o = Some("Foobar");
/// let s = fn_block!{{
///   let foo = o?.get(0..3);
///   Some(foo?.to_lowercase())
/// }};
/// assert_eq!("foo", s.unwrap());
/// ```
///
/// # Example with return type:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// use std::str::from_utf8;
/// use std::error::Error;
/// struct ConvertErr();
/// impl <T: Error> From<T> for ConvertErr {
/// 	fn from(_: T) -> ConvertErr {ConvertErr()}
/// }
/// let s : &[u8] = &[0x0020,0x0034,0x0032];
/// let res_int = fn_expr!{ Result<u32,ConvertErr>:
/// 	from_utf8(s)?.trim().parse::<u32>()?.into_ok()
/// }.unwrap_or(0u32);
/// assert_eq!(res_int, 42);
/// ```
///
/// Note that the example use the trait [`IntoOk`],
/// defined in this crate.
///
/// [`IntoOk`]: trait.IntoOk.html
#[macro_export]
macro_rules! fn_expr {
    ($return_type:ty : $body:expr) => {
        (|| -> $return_type { $body })()
    };
    ($body:expr) => {
        (|| $body)()
    };
}

/// This macro wraps a given rust code expression into a closure and
/// directly calls the closure. The result type of the expression is expected
/// to be an "unwrapped" sucess value (not a `Result` type).
/// The error case (a failing case of a `?` operator) *must* be handled
/// (and recovered to a success type value) by a following `=> catch` block.
/// - *Note 1*: Under the hood the result value of the expression will automatically wrapped
///   into a `Result::Ok`, which is different from how the `fn_expr` and `fn_block` macros work!
/// - *Note 2*: This macro is an unstable API to make use of it, enable the crate feature "unproven".
///
/// # Example:
/// ```
/// # #[macro_use]
/// # use fn_block::*;
/// use std::num::ParseIntError;
/// use std::str::Utf8Error;
/// use std::str::from_utf8;
///
/// enum ConvertErr {
/// 	StrParseErr,
/// 	IntParseErr
/// }
///
/// impl From<Utf8Error> for ConvertErr {
/// 	fn from(_: Utf8Error) -> ConvertErr {
/// 		ConvertErr::StrParseErr
/// 	}
/// }
/// impl From<ParseIntError> for ConvertErr {
/// 	fn from(_: ParseIntError) -> ConvertErr {
/// 		ConvertErr::IntParseErr
/// 	}
/// }
///
/// let s: &[u8] = &[0x0020, 0x0034, 0x0032];
/// let i = fn_try! {
///     from_utf8(s)?.trim().parse::<u32>()?
/// 	=> catch {
/// 		ConvertErr::StrParseErr => 0u32,
/// 		ConvertErr::IntParseErr => u32::max_value()
/// 	}
/// };
/// assert_eq!(42, i);
/// ```
/// Depending on the error type used in the catch block the type is inferred
/// which error type the errors raised in the closure are converted into. This
/// is part of the `?` operator semantics. Note that this also implies, that
/// the `_` pattern cannot be used as the only catch pattern for the error,
/// since in this case the error type cannot be inferred.
///
/// It is advised to use a crate like [`failure`] for error management/conversion.
///
/// # Note of Caution
///
/// Note that this API may be subject of change! The names may change, and the
/// automatic wrapping of result value may disappear. This functionality may be
/// controversial and feedback is welcome if this functionality should stay.
///
/// # Internal workings
///
/// The returned `Result` from the closure will be matched. If an `Ok` is
/// wrapped return value will be returned from the `fn_try`. If the returned
/// result wrapps an error, the error type must be handled by the `=> catch` block
/// Following the expression given by the user. This is basically a match block where
/// the user has to define recovery cases matching error types to the success return type.
///
/// # Design descisions
///
/// The name of the macro and the `=> catch` block are chosen to be similar to the ones
/// chosen for [RFC 2388] and should still work with the "Rust 2018 Edition".
/// Unfortunately the macro name may confuse users of the deprecated `try!`
/// macro, but it looks similar to `try`/`catch` blocks in other languages. The automatic
/// wrapping of the sucessful result value into a `Result::Ok` may also be controversial
/// and even the author is not entirely sure if this is the best way to model the API.
/// However, [RFC 2388] already seems to settle on the automatic wrapping and the resulting
/// code may look more familiar to people comming from other languages.
///
/// [RFC 2388]: https://rust-lang.github.io/rfcs/2388-try-expr.html
/// [`failure`]: https://crates.io/crates/failure
#[macro_export]
#[cfg(feature = "unproven")]
macro_rules! fn_try {
	($body:expr => catch {
		$($err_pat:pat => $pat_bod:expr),+
	}) => {
		match (|| { Ok($body) })() {
			Ok(v) => v,
			Err(e) => match e {
				$($err_pat => $pat_bod),+
			}
		}
	};
}

///////////////////////
// Trait definitions //
///////////////////////

/// This trait, which is implemented for all sized types,
/// provides the method `into_some`, which moves the
/// value on which it is called into an `Optional::Some`.
/// This is particularly useful when having to wrap a value into
/// a `Some` at the end of a call chain.
///
/// # Example:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// let o : Option<String> = "foo bar ".trim().to_uppercase().into_some();
/// assert_eq!("FOO BAR", o.unwrap());
/// ```
///
/// This can e.g. be used inside of
/// an expression wrapped in a [`fn_expr!`] or [`fn_block!`] macro.
///
/// # Example using `fn_expr!`:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// let o = Some("Foobar");
/// let s = fn_expr!{ o?.get(0..3)?.to_lowercase().into_some() };
/// assert_eq!("foo", s.unwrap());
/// ```
///
/// [`fn_expr!`]: macro.fn_expr.html
/// [`fn_block!`]: macro.fn_block.html
pub trait IntoSome: Sized {
    fn into_some(self) -> Option<Self>;
}

/// Implementration of trait `IntoSome` for
/// all sized types.
///
/// # Example
///
/// ```rust
/// # use fn_block::IntoSome;
/// let five = 5.into_some();
/// assert_eq!(Some(5), five);
/// ```
///
impl<T> IntoSome for T {
    /// This method moves `self` into a `Some` and returns it.
    fn into_some(self) -> Option<Self> {
        Some(self)
    }
}

/// This trait, which is implemented for all sized types,
/// provides the method `into_ok`, which moves the
/// value on which it is called into an `Result::Ok`.
/// This is particularly useful when having to wrap a value into
/// an `Ok` at the end of a call chain.
///
/// # Example:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// let res : Result<String,()> = "foo bar ".trim().to_uppercase().into_ok();
/// assert_eq!("FOO BAR", res.unwrap());
/// ```
///
/// This can e.g. be used inside of
/// an expression wrapped in a [`fn_expr!`] or [`fn_block!`] macro.
///
/// # Example using `fn_expr!`:
///
/// ```rust
/// # #[macro_use]
/// # use fn_block::*;
/// use std::str::from_utf8;
/// use std::error::Error;
/// struct ConvertErr();
/// impl <T: Error> From<T> for ConvertErr {
/// 	fn from(_: T) -> ConvertErr {ConvertErr()}
/// }
/// let s : &[u8] = &[0x0020,0x0034,0x0032];
/// let res_int = fn_expr!{ Result<u32,ConvertErr>:
/// 	from_utf8(s)?.trim().parse::<u32>()?.into_ok()
/// }.unwrap_or(0u32);
/// assert_eq!(res_int, 42);
/// ```
///
/// [`fn_expr!`]: macro.fn_expr.html
/// [`fn_block!`]: macro.fn_block.html
pub trait IntoOk<E>: Sized {
    /// This method moves `self` into an `Ok` and returns it.
    fn into_ok(self) -> Result<Self, E>;
}

/// Implementration of trait `IntoOk` for
/// all sized types.
///
/// # Example
///
/// ```rust
/// # use fn_block::IntoOk;
/// let five : Result<u32,()> = 5.into_ok();
/// assert_eq!(Ok(5), five);
/// ```
///
impl<T, E> IntoOk<E> for T {
    fn into_ok(self) -> Result<Self, E> {
        Ok(self)
    }
}

#[macro_use]
#[cfg(test)]
mod tests;
